pkg/client/cache/delta_fifo.go

```diff
@@ -0,0 +1,391 @@
+// after calling this function. f's queue is reset, too; upon return, it
+// will contain the items in the map, in no particular order.
+func (f *DeltaFIFO) Replace(list []interface{}) error {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	for _, item := range list {
+		if err := f.queueActionLocked(Sync, item); err != nil {
+			return fmt.Errorf("couldn't enqueue object: %v", err)
+		}
+	}
+	if f.knownObjectKeys == nil {
+		return nil
+	}
+
+	keySet := make(util.StringSet, len(list))
+	for _, item := range list {
+		key, err := f.KeyOf(item)
+		if err != nil {
+			return fmt.Errorf("couldn't create key for object: %#v", err)
+		}
+		keySet.Insert(key)
+	}
+
+	// Detect deletions not already in the queue.
+	knownKeys := f.knownObjectKeys.ListKeys()
+	for _, k := range knownKeys {
+		if _, exists := keySet[k]; exists {
+			continue
+		}
+
+		// This key isn't in the complete set we got, so it must have been deleted.
+		if d, exists := f.items[k]; exists {
+			// Don't issue a delete delta if we have one enqueued as the most
+			// recent delta.
+			if d.Newest().Type == Deleted {
+				continue
+			}
+		}
+		if err := f.queueActionLocked(Deleted, DeletedFinalStateUnknown{k}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// NewDeltaFIFO returns a Store which can be used process changes to items.
+//
+// keyFunc is used to figure out what key an object should have. (It's
+// exposed in the returned DeltaFIFO's KeyOf() method, with bonus features.)
+//
+// 'compressor' may compress as many or as few items as it wants
+// (including returning an empty slice), but it should do what it
+// does quickly since it is called while the queue is locked.
+// 'compressor' may be nil if you don't want any delta compression.
+//
+// 'keyLister' is expected to return a list of keys that the consumer of
+// this queue "knows about". It is used to decide which items are missing
+// when Replace() is called; 'Deleted' deltas are produced for these items.
+// It may be nil if you don't need to detect all deletions.
+// TODO: consider merging keyLister with this object, tracking a list of
+//       "known" keys when Pop() is called. Have to think about how that
+//       affects error retrying.
+func NewDeltaFIFO(keyFunc KeyFunc, compressor DeltaCompressor, knownObjectKeys KeyLister) *DeltaFIFO {
+	f := &DeltaFIFO{
+		items:           map[string]Deltas{},
+		queue:           []string{},
+		keyFunc:         keyFunc,
+		deltaCompressor: compressor,
+		knownObjectKeys: knownObjectKeys,
+	}
+	f.cond.L = &f.lock
+	return f
+}
+
+// A KeyLister is anything that knows how to list its keys.
+type KeyLister interface {
+	ListKeys() []string
+}
+
+// KeyListerFunc adapts a raw function to be a KeyLister.
+type KeyListerFunc func() []string
+
+// ListKeys just calls kl.
+func (kl KeyListerFunc) ListKeys() []string {
+	return kl()
+}
+
+// DeltaCompressor is an algorithm that removes redundant changes.
+type DeltaCompressor interface {
```

Comment:
But with the function typedef you can just pass foo.Function instead of passing the full foo, I think that is actually much clearer.  There's no interface that your object implements.  It just has a function that implements that interface.